tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SNS/ContextPropagation.cs

```diff
@@ -27,14 +28,34 @@ private static void Inject<TMessageRequest>(SpanContext context, IDictionary mes
             sb.Append('}');
 
             var resultString = Util.StringBuilderCache.GetStringAndRelease(sb);
-            byte[] bytes = Encoding.UTF8.GetBytes(resultString);
-            MemoryStream stream = new MemoryStream(bytes);
+            var bytes = Encoding.UTF8.GetBytes(resultString);
+            var stream = new MemoryStream(bytes);
             messageAttributes[SnsKey] = CachedMessageHeadersHelper<TMessageRequest>.CreateMessageAttributeValue(stream);
         }
 
-        public static void InjectHeadersIntoMessage<TMessageRequest>(IContainsMessageAttributes carrier, SpanContext spanContext)
+        public static void InjectHeadersIntoBatch<TBatchRequest>(IPublishBatchRequest request, SpanContext context)
         {
-            // add distributed tracing headers to the message
+            // Skip adding Trace Context if entries don't exist or empty.
+            if (request.PublishBatchRequestEntries is not { Count: > 0 })
+            {
+                return;
+            }
+
+            var publishBatchRequestEntry = request.PublishBatchRequestEntries[0].DuckCast<IContainsMessageAttributes>();
+
+            InjectHeadersIntoMessage<TBatchRequest>(publishBatchRequestEntry, context);
+        }
+
+        public static void InjectHeadersIntoMessage<TMessageRequest>(IContainsMessageAttributes carrier, SpanContext context)
```

Comment:
We should use `where` to avoid boxing

```suggestion
        public static void InjectHeadersIntoMessage<TMessageRequest>(TMessageRequest carrier, SpanContext context)
            where TMessageRequest : IContainsMessageAttributes
```

Reply:
Resolving due to [this comment](https://github.com/DataDog/dd-trace-dotnet/pull/4712#issuecomment-1762212524).